pkg/operator/ceph/object/zone/controller.go

```diff
@@ -195,6 +200,55 @@ func (r *ReconcileObjectZone) reconcile(request reconcile.Request) (reconcile.Re
 		return reconcileResponse, *cephObjectZone, err
 	}
 
+	// DELETE: the CR was deleted
+	if !cephObjectZone.GetDeletionTimestamp().IsZero() {
+		logger.Debugf("deleting zone CR %q", cephObjectZone.Name)
+		objContext := object.NewContext(r.context, r.clusterInfo, cephObjectZone.Name)
+		objContext.Realm = realmName
+		objContext.ZoneGroup = cephObjectZone.Spec.ZoneGroup
+		objContext.Zone = cephObjectZone.Name
+		zonePresent, err := object.CheckIfZonePresentinZoneGroup(objContext)
+		if err != nil {
+			return reconcile.Result{}, *cephObjectZone, err
+		}
+		if zonePresent {
+			deps, err := CephObjectZoneDependentStores(r.context, r.clusterInfo, cephObjectZone, objContext)
+			if err != nil {
+				return reconcile.Result{}, *cephObjectZone, err
+			}
+			if !deps.Empty() {
+				err := reporting.ReportDeletionBlockedDueToDependents(r.opManagerContext, logger, r.client, cephObjectZone, deps)
+				return opcontroller.WaitForRequeueIfFinalizerBlocked, *cephObjectZone, err
+			}
+			reporting.ReportDeletionNotBlockedDueToDependents(r.opManagerContext, logger, r.client, r.recorder, cephObjectZone)
+			zoneIsMaster, err := object.CheckZoneIsMaster(objContext)
+			if err != nil {
+				return reconcile.Result{}, *cephObjectZone, err
+			}
+			if !zoneIsMaster {
+				err = r.removeZoneFromZonegroup(objContext)
+				if err != nil {
+					return reconcile.Result{}, *cephObjectZone, err
+				}
+			}
+			if !cephObjectZone.Spec.PreservePoolsOnDelete {
+				err = r.deletePoolsAndZone(objContext, cephObjectZone, realmName)
+				if err != nil {
+					return r.setFailedStatus(k8sutil.ObservedGenerationNotAvailable, cephObjectZone, request.NamespacedName, "failed to delete ceph zone", err)
+				}
+			} else {
+				logger.Infof("PreservePoolsOnDelete is set in object zone %s. Pools and Zone are not deleted", objContext.Name)
+			}
+			// Remove finalizer
+			err = opcontroller.RemoveFinalizer(r.opManagerContext, r.client, cephObjectZone)
+			if err != nil {
+				return reconcile.Result{}, *cephObjectZone, errors.Wrap(err, "failed to remove finalizer")
+			}
+			// Return and do not requeue. Successful deletion.
+			return reconcile.Result{}, *cephObjectZone, nil
```

Comment:
All this text inside the main reconcile function makes it longer and harder to follow. Break this off into a function.